Source/Scene/Cesium3DTileset.js

```diff
@@ -1246,6 +1332,101 @@ define([
+
+        var length = finalQueue.length;
+        for (var i = 0; i < length; ++i) {
+            var original = finalQueue.get(i);
+            var tile = original;
+            // traverse up the tree to find a ready ancestor
+            if (tile.hasContent || tile.hasTilesetContent) {  // could be Empty3DTileContent
+                while (defined(tile) && !(tile.hasContent && tile.contentReady)) {
+                    if (!tile.contentReady) {
+                        tileset._hasMixedContent = true;
+                    }
+                    tile = tile.parent;
+                }
+            }
+
+            if (defined(tile)) {
+                if (!tile.selected) {
+                    tile._finalResolution = (tile === original || tile.refine === Cesium3DTileRefine.ADD);
+                    tile.selected = true;
+                    tile._selectedFrame = frameNumber;
+                    selectionQueue.push(tile);
+                }
+            } else {
+                // if no ancestors are ready, traverse down and select ready tiles to minimize empty regions
+                descendantStack.push(original);
+                while (descendantStack.length > 0) {
+                    tile = descendantStack.pop();
+                    var children = tile.children;
+                    var childrenLength = children.length;
+                    for (var j = 0; j < childrenLength; ++j) {
+                        var child = children[j];
+                        if (child.contentReady) {
+                            if (!child.selected) {
+                                child._finalResolution = true;
+                                child.selected = true;
+                                child._selectedFrame = frameNumber;
+                                touch(tileset, child, outOfCore);
+                                selectionQueue.push(child);
+                            }
+                        }
+                        if (child._depth - original._depth < 2) { // prevent traversing too far
+                            if (!child.contentReady || child.refine === Cesium3DTileRefine.ADD) {
+                                descendantStack.push(child);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        selectionQueue.trim();
+    }
+
+    function markTilesAsFinal(tiles) {
+        var length = tiles.length;
+        tiles = tiles.internalData;
+        var i;
+        for (i = 0; i < length; ++i) {
+            tiles[i]._finalResolution = true;
+        }
+
+        for (i = 0; i < length; ++i) {
+            var parent = tiles[i].parent;
+            while (defined(parent)) {
+                // tiles using additive refinement are always final
+                parent._finalResolution = (parent.refine === Cesium3DTileRefine.ADD);
+                parent = parent.parent;
+            }
+        }
+    }
+
+    var selectionState = {
```

Comment:
If you're sure there only needs to be one of these - not one per tileset - rename it to `selectionStateScratch`, define it above its first use, and access it via a closure instead of passing it around.